### PR TITLE
Implementation of the Core for VPDU

### PR DESCRIPTION
**Pull request by idjaw**
This implementation provides the core implementation layer linking the pdu to
the driver

### CODE REVIEW

**Comment by mat128** [comment]
Please squash the branch prior to opening the pull request :)

Also, I think the get_power_state was wrongfully merged, because it's already part of master and I think your issue is related to the `LIBVIRT_TEST_PROVIDER = 'test+unix:///default'` change that was part of an older revision.

**Comment by mat128** [comment]
:+1:

File: virtualpdu/core.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+REBOOT = 'REBOOT'
+
+
+class Core(object):
+    def __init__(self, driver, mapping):
+        self.driver = driver
+        self.mapping = mapping
+        self.logger = logging.getLogger(__name__)
+
+    def _get_server_name(self, pdu_name,
```
Comment (fbouliane) [inline virtualpdu/core.py]: This method should go in the bottom
after pdu_outlet_state_changed

File: virtualpdu/core.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+                         outlet_number):
```
Comment (fbouliane) [inline virtualpdu/core.py]: unnecessary line break

File: virtualpdu/core.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        return self.mapping[(pdu_name, outlet_number)]
+
+    def pdu_outlet_state_changed(self, name, outlet_number, state):
+        self.logger.warning(
```
Comment (fbouliane) [inline virtualpdu/core.py]: self.logger.debug ?

File: virtualpdu/core.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            self.logger.warning("Found server '{0}'".format(server_name))
+        except KeyError:
+            return
+
+        if state == POWER_ON:
+            self.driver.power_on(server_name)
+        elif state == POWER_OFF:
+            self.driver.power_off(server_name)
+        elif state == REBOOT:
+            self.driver.power_off(server_name)
```
Comment (fbouliane) [inline virtualpdu/core.py:43]: future improvement, add LibvirtDriver:reboot and use [libvirt's reboot](https://libvirt.org/sources/java/javadoc/org/libvirt/Domain.html#reboot%28int%29)

File: virtualpdu/tests/integration/test_core_integration.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+class TestCoreIntegration(base.TestCase):
+    def tearDown(self):
+        self.pdu_test_harness.stop()
+        super(TestCoreIntegration, self).tearDown()
+
+    def test_pdu_outlet_state_changed_on_power_off(self):
+        mapping = {
+            ('my_pdu', 5): 'test'
+        }
+        self.driver = libvirt_driver.KeepaliveLibvirtDriver('test:///default')
```
Comment (fbouliane) [inline virtualpdu/tests/integration/test_core_integration.py]: No need for putting things in self if not reused.

File: virtualpdu/core.py (hunk 4 above)
Comment (mat128) [reply virtualpdu/core.py:43]: nope.

Comment (mat128) [reply virtualpdu/core.py:43]: I'd like the core to control what a reboot is. In the future, we might have "delayedReboot" because APC has it.

Comment (mat128) [reply virtualpdu/core.py:43]: I'd rather not. The core (or even the PDU) should control what a reboot is. APC offers a "delayedReboot" which is basically off, sleep(n), on. That shouldn't be in libvirt's hands.

Comment (fbouliane) [reply virtualpdu/core.py:43]: ok, keep as is :+1:

**Comment by fbouliane** [comment]
LGTM, other than the last small comment :+1:

**Pull request merged**